### DISCUSSION TITLE
Interact with incoming Websocket messages from within the App Studio

### DISCUSSION THREAD

**Discussion by amo42** [directus/directus#18953] [Feature Requests]
### Summary

As the websockets are now released it would be handy to listen the websocket events from within the app studio. Especially  within flows, the websocket message events could be handy to track the connections and trigger any kind of automation.

### Basic Example

As an example, think of the provided Real Time Chat example. Everytime, a new connection is established, a Flow could be triggered to store some kind of end user details  such as and GUID. having this, anyone els joining the chat later, would be able to see the currently connected users. 
Same goes for closing the connection, once the connection is either closed by timeout or by user, there is no chance to notify the backend "hey I'm out". If the close event can be handled by a Flow, any logic for the closed connection can be build with the flow e.g. update a "session" collection to notify all remaining subscripted clients.

### Motivation

To have for instance a list of active subscription in flavor of a collection, maintained by the described Flows from the example, would allow you to not only broadcast, but also handle closed connections and interact with specific end clients if the subscript to only there relevant details.

### Detailed Design

As the events were internally available, they should be accessible  as trigger for Flows as well. Similar to the CRUD of a collection the basic events just as 

- new connection
- incoming message
- outgoing message
- connection closed

Should be usable for a Flow.

### Must Have:
- Public websocket events to use them within flows
- Access the payload/message details

### Should Have:
- Connection details, if possible
- Possibility to suppress events   

### Could Have:
- Prevent loops or at least detect and warn 

### Won't Have:
- Any further logic, as this would be up to the user what to do with this events

### Drawbacks

As it is just about to publish/listening to existing events, it should be not break anything. 

### Alternatives

An extension that just handover the event data to a collection, which could trigger a Flow afterwards

### Adoption Strategy

It would offer even more possibilities to the new websocket feature.

**Comment by rijkvanzanten** [comment]
I'm pretty sure WebSockets already emit events under the hood, so they might theoretically already be available to extensions and flows. @br41nslug thoughts? If that's the case, it might be as easy as adding them to the dropdown of events and document them in the extensions reference for hooks 🙂

**Reply by amo42** [reply]
I was hoping for a "low hanging fruit" regardless of the effort, I think it could offer a lot of benefits anyway. But not sure if
 I might miss some technical limitation that would increase the complexity.

**Reply by rijkvanzanten** [reply]
This is a bit of a stretch, but worth a shot: Try setting up a new `filter` flow, use Edit Raw Value on the filter event, and set it to `websocket.message`. I think it might already work!

**Reply by amo42** [reply]
Oh and it works like a charm, I can confirm they are usable this way
<img width="1002" alt="image" src="https://github.com/directus/directus/assets/12328155/08da2510-a15b-4e5c-989d-7611f7b8c7e6">

**Reply by br41nslug** [reply]
Bit late to the party but yeah the websocket hook events are available and usable (but not selectable). Beware tho that these `websocket.*` do fire on each instances if you have a horizontally scaled setup.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
